### DISCUSSION TITLE
How to create Informer w/ server-side filtering rules?

### DISCUSSION THREAD

**Discussion by pan3793** [fabric8io/kubernetes-client#5011] [Q&A] [answered]
Following the example, I found that the `podInformer` will poll all existing pods' events on starting, which may cause pressure for both the api-server and client.

https://github.com/fabric8io/kubernetes-client/blob/master/kubernetes-examples/src/main/java/io/fabric8/kubernetes/examples/SharedInformerExample.java

And after some investigation, I found an API, but seems it does filtering on the client side.

https://github.com/fabric8io/kubernetes-client/blob/master/kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/dsl/Informable.java#L135

So my question is how to create an Informer w/ server-side filtering rules?

Looks like the Go client has such an API, any equivalent on Java side?
https://github.com/kubernetes/client-go/blob/master/informers/factory.go#L115

**Comment by rohanKanojia** [comment]
> And after some investigation, I found an API, but seems it does filtering on the client side.

Could you please elaborate a bit on this point?

**Reply by pan3793** [reply]
The Java Doc says

> The predicate will be tested against the state of the underlying informer store on every event.

And the implementation proves that, obviously, the predication is tested on the client side

https://github.com/fabric8io/kubernetes-client/blob/master/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/BaseOperation.java#L914

**Comment by shawkins** [comment, marked as answer]
There are different levels to the informer related apis.   At the core you are in control over how much the informer watches.

pods().inAnyNamespace().inform(...) - watches all pods

The inform methods however are available where a single item or a list of items is available:

pods().inNamespace("ns").withLabel("some", "label").inform(...) - watches pods with the given label only in namespace ns.

The resulting SharedIndexInformer you can use however your application needs - via its cache and event handlers.

A method like informOnCondition is a convenience method to provide the evaluation of a predicate over an informer - you use this or methods like waitUntilReady when you don't generally need to track a given resource, but just need to be notified or block until a particular state is met.  As far as I know there is no ability to have that high level of a predicate evaluated server-side.

**Reply by pan3793** [reply]
> `pods().inNamespace("ns").withLabel("some", "label").inform(...)`

So, this is kind of server-side filtering w/ rule `namespace = "ns" & label["some"] = "label"`?

**Reply by shawkins** [reply]
Yes, anything supplied before the inform call in the dsl is used to form the appropriate urls for the list / watch operations.

**Reply by pan3793** [reply]
Got it, thanks
